src/duckdb.ts

```diff
@@ -0,0 +1,97 @@
+import {existsSync} from "node:fs";
+import {mkdir, writeFile} from "node:fs/promises";
+import {dirname, join} from "node:path/posix";
+import {cross} from "d3-array";
+import type {DuckDBConfig} from "./config.js";
+import {faint} from "./tty.js";
+
+const downloadRequests = new Map<string, Promise<string>>();
+
+export async function duckDBManifest(
+  duckdb: DuckDBConfig,
+  {root, log, aliases}: {root: string; log?: boolean; aliases?: Map<string, string>}
+) {
+  return {
+    bundles: duckdb.bundles,
+    extensions: await Promise.all(
+      cross(duckdb.bundles, duckdb.install).map(async ([p, name]) => {
+        let ext = await resolveDuckDBExtension(root, p, duckdb.source[name], name);
+        if (aliases?.has(ext)) ext = aliases.get(ext)!;
+        return [
+          name,
+          {
+            ref: dirname(dirname(dirname(ext))),
+            load: duckdb.load.includes(name),
+            bundle: p
+          }
+        ];
+      })
+    ),
+    log
+  };
+}
+
+/**
+ * Given a duckdb configuration and an extension name such as "parquet", saves
+ * the binary to _duckdb/{hash}/v1.1.1/wasm_{p}/parquet.duckdb_extension.wasm
+ * for every supported platform p ("eh" and "mvp"), and returns a content-hashed
+ * reference (_duckdb/{hash}) to use in the corresponding DuckDB INSTALL
+ * statement. The repo is structured as required by DuckDB with:
+ * ${repo}/v1.1.1/wasm_{p}/${name}.duckdb_extension.wasm
+ */
+export async function resolveDuckDBExtension(root: string, p: string, repo: string, name: string): Promise<string> {
```

Comment:
This is downloading the extension greedily during resolution, but it looks to me like we could defer that to later, and that we know synchronously what the URL of the extension will be. If we can make DuckDB extension resolution synchronous, that might be nice for simplicity and performance (only having the Preview server fetch the file lazily when requested). But this might be related to my [other question](https://github.com/observablehq/framework/pull/1734/files#r1807904150) on the `v1.1.1`; I’m guessing we need to resolve the latest version of the extension, in which case this still needs to be asynchronous.